src/include/stat.h

```diff
@@ -6,34 +6,160 @@
+ * that, but not all (regardless, a system call every time increment a stats
+ * counter is far too expensive).
+ *
+ * Our second-best option is to use the thread ID. Unfortunately, there is no
+ * portable way to obtain a unique thread ID that is a small-enough number to
+ * be used as an array index (portable thread IDs are usually a pointer or an
+ * opaque chunk, not a simple integer).
+ *
+ * Our solution is to use the session ID; there is normally a session per thread
+ * and the session ID is a small, monotonically increasing number.
  */
-#define	WT_STAT(stats, fld)						\
-	((stats)->fld.v)
-#define	WT_STAT_ATOMIC_DECRV(stats, fld, value) do {			\
-	(void)WT_ATOMIC_SUB8(WT_STAT(stats, fld), (value));		\
-} while (0)
-#define	WT_STAT_ATOMIC_DECR(stats, fld) WT_STAT_ATOMIC_DECRV(stats, fld, 1)
-#define	WT_STAT_ATOMIC_INCRV(stats, fld, value) do {			\
-	(void)WT_ATOMIC_ADD8(WT_STAT(stats, fld), (value));		\
-} while (0)
-#define	WT_STAT_ATOMIC_INCR(stats, fld) WT_STAT_ATOMIC_INCRV(stats, fld, 1)
-#define	WT_STAT_DECRV(stats, fld, value) do {				\
-	(stats)->fld.v -= (value);					\
-} while (0)
-#define	WT_STAT_DECR(stats, fld) WT_STAT_DECRV(stats, fld, 1)
-#define	WT_STAT_INCRV(stats, fld, value) do {				\
-	(stats)->fld.v += (value);					\
+#define	WT_STATS_SLOT_ID(session)					\
+	((session)->id) % WT_COUNTER_SLOTS
+
+/*
+ * Set all the values in the array counter slots to zero. We do more work than
+ * necessary by using memset, because we are clearing both the values and the
+ * padding. However, resetting the counters is not a common case operation, so
+ * we use memset for compactness.
+ */
+#define	WT_STAT_ALL_RESET(stats, fld) do {				\
+	memset((stats)->fld.array_v, 0,					\
+	    sizeof(WT_PADDED_COUNTER) * WT_COUNTER_SLOTS);		\
 } while (0)
-#define	WT_STAT_INCR(stats, fld) WT_STAT_INCRV(stats, fld, 1)
-#define	WT_STAT_SET(stats, fld, value) do {				\
-	(stats)->fld.v = (uint64_t)(value);				\
+
+/*
+ * Aggregate the counter values from all slots into the "master" value "v",
+ * return v.
+ */
+static inline uint64_t
+__wt_stats_aggregate_and_return(struct __wt_stats *stats)
+{
+	int64_t aggr_v;
+	int i;
+
+	for (aggr_v = 0, i = 0; i < WT_COUNTER_SLOTS; i++)
+		aggr_v += stats->array_v[i].v;
+
+	/*
+	 * This can race. However, any implementation with a single value can
+	 * race as well, different threads could set the same counter value
+	 * simultaneously. While we are making races more likely, we are not
+	 * fundamentally weakening the isolation semantics found in updating a
+	 * single value.
+	 *
+	 * Additionally, the aggregation can go negative (imagine a thread
+	 * incrementing a value after aggregation has passed its slot and a
+	 * second thread decrementing a value before aggregation has reached
+	 * its slot). Limit the return to 0, negative numbers will just look
+	 * really, really large.
+	 */
+	if (aggr_v < 0)
+		aggr_v = 0;
+	return ((uint64_t)aggr_v);
+}
```

Comment:
@fedorova, I think we can sum to a negative value, so I think we should bound the return to 0. Does that look correct to you?